InnerEye/ML/SSL/lightning_modules/ssl_online_evaluator.py

```diff
@@ -0,0 +1,139 @@
+        """
+        Forward pass and MLP loss computation for the linear head only. Representations from the encoder are frozen and
+        detach from computation graph for this loss computation.
+        Returns cross-entropy loss for the input batch.
+        """
+        batch = batch[SSLDataModuleType.LINEAR_HEAD] if isinstance(batch, dict) else batch
+        x, y = self.to_device(batch, pl_module.device)
+        with torch.no_grad():
+            representations = self.get_representations(pl_module, x)
+        representations = representations.detach()
+        assert isinstance(pl_module.non_linear_evaluator, torch.nn.Module)
+
+        # Run the linear-head with SSL embeddings.
+        mlp_preds = pl_module.non_linear_evaluator(representations)
+        weights = None if self.class_weights is None else self.class_weights.to(device=pl_module.device)
+        mlp_loss = F.cross_entropy(mlp_preds, y, weight=weights)
+
+        with torch.no_grad():
+            posteriors = F.softmax(mlp_preds, dim=-1)
+            for metric in (self.train_metrics if is_training else self.val_metrics):
+                metric(posteriors, y)  # type: ignore
+
+        return mlp_loss
+
+    def on_validation_batch_end(self, trainer: pl.Trainer,
+                                pl_module: pl.LightningModule,
+                                outputs: Any,
+                                batch: BatchType,
+                                batch_idx: int,
+                                dataloader_idx: int) -> None:  # type: ignore
+        """
+        Get and log validation metrics.
+        """
+        ids_linear_head = tuple(batch[SSLDataModuleType.LINEAR_HEAD][0].tolist())
+        if ids_linear_head not in self.visited_ids:
+            self.visited_ids.add(ids_linear_head)
+            loss = self.shared_step(batch, pl_module, is_training=False)
+            pl_module.log('ssl_online_evaluator/val/loss', loss, on_step=False, on_epoch=True, sync_dist=False)
+            for metric in self.val_metrics:
+                pl_module.log(f"ssl_online_evaluator/val/{metric.name}", metric, on_epoch=True,
+                              on_step=False)  # type: ignore
+
+    def on_train_batch_end(self, trainer, pl_module, outputs, batch, batch_idx, dataloader_idx) -> None:  # type: ignore
+        """
+        Get and log training metrics, perform network update.
+        """
+        ids_linear_head = tuple(batch[SSLDataModuleType.LINEAR_HEAD][0].tolist())
+        if ids_linear_head not in self.visited_ids:
+            self.visited_ids.add(ids_linear_head)
+            loss = self.shared_step(batch, pl_module, is_training=True)
+            loss.backward()
+            self.optimizer.step()
+            self.optimizer.zero_grad()
+            self.training_step += 1
+
+            # log metrics
+            pl_module.log('ssl_online_evaluator/train/loss', loss)
```

Comment:
did they fix this bug in training step logging in pl_module callbacks?

Reply:
It does log in AML for sure, i haven't checked in tensorflow

Reply:
Yep they fixed it